src/Services/Virtualization/TokenAuthentication.cs

```diff
@@ -0,0 +1,352 @@
+                };
+
+                var tokenManager = new TokenManager(SecurityKey, tokenHandler, generateTokenParameter);
+
+                switch (tokenAction)
+                {
+                    case TokenAction.TokenLogin:
+                        endRequest = true;
+                        TokenLogin(basicAuthenticated, validFrom, tokenManager, context);
+                        break;
+                    case TokenAction.TokenLogout:
+                        endRequest = true;
+                        TokenLogout(tokenHeadAndPayload, tokenManager, context);
+                        break;
+                    case TokenAction.TokenAccess:
+                        TokenAccess(tokenHeadAndPayload, tokenManager, context);
+                        break;
+                    case TokenAction.TokenRefresh:
+                        endRequest = true;
+                        TokenRefresh(tokenHeadAndPayload, validFrom, tokenManager, context);
+                        break;
+                }
+            }
+            catch (Exception ex)
+            {
+                SnLog.WriteException(ex);
+                if (endRequest)
+                {
+                    context.Response.StatusCode = HttpResponseStatusCode.Unauthorized;
+                }
+                else
+                {
+                    context.User = AuthenticationHelper.GetVisitorPrincipal();
+                }
+            }
+            finally
+            {
+                if (endRequest)
+                {
+                    context.Response.Flush();
+                    if (application.Context != null)
+                    {
+                        application.CompleteRequest();
+                    }
+                }
+            }
+        }
+
+        private void TokenLogin(bool basicAuthenticated, DateTime validFrom, TokenManager tokenManager, HttpContextBase context)
+        {
+            if (!basicAuthenticated)
+            {
+                throw new AuthenticationException("Missing basic authentication.");
+            }
+            // user has just authenticated by basic auth, so let's emit a set of tokens and cookies in response 
+            var userName = context.User.Identity.Name;
+            var roleName = String.Empty;
+
+            // emit both access and refresh token and cookie 
+            EmitTokensAndCookies(context, tokenManager, validFrom, userName, roleName, true);
+            context.Response.StatusCode = HttpResponseStatusCode.Ok;
+        }
+
+        private void TokenLogout(string accessHeadAndPayload, TokenManager tokenManager, HttpContextBase context)
+        {
+            if (!String.IsNullOrWhiteSpace(accessHeadAndPayload))
+            {
+                var authCookie = CookieHelper.GetCookie(context.Request, AccessSignatureCookieName);
+                if (authCookie == null)
+                {
+                    throw new UnauthorizedAccessException("Missing access cookie.");
```

Comment:
Is it necessary to have a valid token/cookie, when logging out? The outcome should be the same - I want to be logged out. It is the same as with _delete_ operations: if the content is not there anymore, it is OK, that's what we wanted.
For example why would I receive a 401 if my access token expired in the meantime. Who cares, I want to be logged out.